### PR TITLE
MB-58034: Custom Date Time parsers not being applied on query object

### PR DESCRIPTION
**Pull request by CascadingRadium**
## Jira

[MB-58034](https://issues.couchbase.com/browse/MB-58034)

## Description

- Add a new query DateRangeStringQuery, which takes in the start and end bounds as strings and then parses them using either the top level QueryDateTimeParser in the top-level bleve package or the DateTimeParser set in the DateRangeStringQuery struct, if present.
-  ParseQuery, which analyzes a marshalled json to identify the query to run, will now default to DateRangeStringQuery instead, which is better since it already has the dates in string format and unmarshalling to DateRangeQuery will always trigger parsing of the date string with Only QueryDateTimeParser, leaving no room for user config.

### CODE REVIEW

File: search/query/date_range.go
Comment (abhinavdangeti) [inline search/query/date_range.go:33]: Not a fan of the "User path" wording - I understand what you are intending it for, but I think the name of the attribute is self-explanatory enough. let's drop the "User Path" verbiage.

File: search/query/date_range.go
Comment (abhinavdangeti) [inline search/query/date_range.go:88]: Don't see a need for this method to be exported.

Also perhaps we should change the signature to this method to  ..

    dateRangeQuery, err := UnmarshalDateRange(input)

Separately can you help clarify the need for this validation here?

File: search/query/date_range.go
Comment (abhinavdangeti) [inline search/query/date_range.go:198]: Hmm, should probably do this string job only when index mapping for this field has store enabled and the query requested for this field ?

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: search/query/date_range.go
Comment (abhinavdangeti) [reply search/query/date_range.go:33]: Also one rule of thumb - do not add anything "couchbase" specific here. Keep things generic enough.

File: search/query/date_range.go
Comment (CascadingRadium) [reply search/query/date_range.go:88]: ack i will not export
signature is the same for all json.Unmarshall calls in parse.go.
So i have kept it consistent

I am basically avoiding the usage of UnmarshalJSON() which is called when json.Unmarshal is called, for a date range query object in the ParseQuery path. By using a custom unmarshaller i can avoid entering the User Path here. Thus i can keep the start and end as raw strings and NOT time objects. 

When the searcher func is called, i have access to the index mapping and can get the layout to parse the raw start and end strings with.

File: search/query/date_range.go
Comment (CascadingRadium) [reply search/query/date_range.go:198]: the object here is a query object right? like this check is just done to check if the unmarshaller used was my custom one (which is the only place where raw start and raw end is actually set) or the default one used in User Path. these 2 params are not exported so bleve only users cant see this

File: search/query/date_range.go
Comment (CascadingRadium) [reply search/query/date_range.go:212]: *bool was the existing method parameter in the code for NewDateRange and NewDateRangeInclusive. So i have maintained the same here as well.

**Review by Thejas-bhat — APPROVED** [review]

**Comment by Thejas-bhat** [comment]
the changes look good to me, however i'd strongly recommend waiting for others' review as well.

File: search/query/date_range_string.go
Comment (Thejas-bhat) [inline search/query/date_range_string.go:1]: oh missed this, as suggested in other PRs as well. this must be 2023.

File: mapping/index.go
Comment (abhinavdangeti) [inline mapping/index.go:22]: Drop this empty line between these imports.

File: search/query/date_range_string.go
Comment (abhinavdangeti) [inline search/query/date_range_string.go:99]: Drop empty line.

**Review by abhinavdangeti — COMMENTED** [review]
Looking good. Couple minor nits, we can merge this after those are addressed.

**Review by abhinavdangeti — APPROVED** [review]

**Pull request merged**
